url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/builder/SerializedObject.java

```diff
@@ -0,0 +1,129 @@
+*/
+
+class SerializedObject
+{
+	/** Do-nothing constructor. */
+	public SerializedObject()	{
+	}
+	
+	/**
+		Deserializes an object from filesystem.
+		@return deserialized object.
+	*/
+	protected Object read(String fileName)	{
+		fileName = makeFilePath(fileName);
+		System.err.println("deserializing object from "+fileName);
+		ObjectInputStream oin = null;
+		try	{
+			FileInputStream in = new FileInputStream(fileName);
+			oin = new ObjectInputStream(in);
+			return oin.readObject();
+		}
+		catch (Exception e)	{
+			System.err.println(e.getMessage());	// tolerate non-existing object
+			return null;
+		}
+		finally	{
+			try	{	oin.close();	}	catch (Exception e)	{}
+		}
+	}
+	
+	/**
+		Serializes an object to filesystem.
+		@return true on success.
+	*/
+	protected boolean write(String fileName, Object o)	{
+		fileName = makeFilePath(fileName);
+		System.err.println("serializing object to "+fileName);
+		ObjectOutputStream oout = null;
+		try	{
+			ensureDirectory(fileName);
+			FileOutputStream out = new FileOutputStream(fileName);
+			oout = new ObjectOutputStream(out);
+			oout.writeObject(o);
+			return true;
+		}
+		catch (IOException e)	{
+			e.printStackTrace();
+			return false;
+		}
+		finally	{
+			try	{	oout.flush(); oout.close();	}	catch (Exception e)	{}
+		}
+	}
+
+
+	/**
+		When syntaxInput is a File, the name of the serialization file is created
+		from its basename (without any extension), else "Unknown" is assumed as basename.
+		The returned name has no path.
+	*/
+	public static String baseNameFromSyntax(Object syntaxInput)	{
+		String name;
+		if (syntaxInput instanceof File)	{
+			File f = (File) syntaxInput;
+			name = f.getName();
+			int i = name.indexOf(".");
+			if (i > 0)
+				name = name.substring(0, i);
+		}
+		else	{
+			name = "Unknown";
+		}
+		return name;
+	}
+
+
+	/** Converts a File, InputStream, Reader, String, StringBuffer, List of Lists or String[][] to a Syntax. */
+	public static Syntax toSyntax(Object syntaxInput)
+		throws Exception
+	{
+		Syntax syntax;
+		if (syntaxInput instanceof Syntax)
+			syntax = (Syntax) syntaxInput;
+		else
+		if (syntaxInput instanceof String[][])
+			syntax = new Syntax((String[][]) syntaxInput);
+		else
+		if (syntaxInput instanceof List)
+			syntax = new Syntax((List)syntaxInput);
+		else
+			syntax = new SyntaxBuilder(syntaxInput).getSyntax();
+			
+		syntax.resolveSingulars();
+		return syntax;
+	}
+	
+	
+	/** Puts the file into ".friware/parsers" directory in "user.home". */
+	protected String makeFilePath(String fileName)	{
+		return ConfigDir.dir()+"parsers"+File.separator+fileName;
+	}
+	
+	/** Creates the directory of passed filename if it does not exist. @return the directory name. */
+	protected String ensureDirectory(String fileName)	{
+		File file = new File(fileName);
+		String dir = file.getParent();
+		
+		if (dir != null)	{
+			File directory = new File(dir);
+			if (directory.exists() == false)
+				directory.mkdirs();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Confirm that `mkdirs` was successful. Handle any failures that happened. If you fail silently, then you might encounter errors that are harder to debug. See the following diffs on GitHub that make this code change: [diff1](https://github.com/qunarcorp/qmq/commit/12f1d9d6a57a02bc2415cb2af9f4388a77b12c02#diff-6b67be6e17f24fdeddb7c725ac8ec5600d6e63be0bdca0963a83d530e6d55e35L36), [diff2](https://github.com/apache/jmeter/commit/1ec61206779b3f84df18516682b4791bf015178f#diff-cd74304763c097fbc085ac2bc952b46ab3843ac825cacacad89e8064296e955aL330)

Learn more: [File.mkdirs in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#mkdirs--)